src/js/url.js

```diff
@@ -0,0 +1,256 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+function Url(url) {
+  this.protocol = null;
+  this.auth = null;
+  this.host = null;
+  this.port = null;
+  this.hostname = null;
+  this.afterHost = null;
+  if (url) {
+    this.parse(url);
+  }
+}
+
+var protocolPattern = /^[a-z0-9.+-]+:/i;
+var portPattern = /:[0-9]*$/;
+
+var hostnameMaxLen = 255;
+var maxPortNumber = 65535;
+
+var supportedProtocol = {
+  'http': true,
+  'http:': true,
+  'https': true,
+  'https:': true,
+};
+
+// The parser
+function urlParse(url) {
+  if (url instanceof Url) {
+    return url;
+  }
+
+  var urlObject = new Url();
+  urlObject.parse(url);
+  return urlObject;
+}
+
+
+Url.prototype.parse = function(url) {
+  if (typeof url !== 'string') {
+    throw new TypeError('ERR_INVALID_ARG_TYPE', 'url', 'string', url);
+  }
+
+  var rest = url;
+
+  // Get the protocol first
+  var proto = protocolPattern.exec(rest);
+  if (proto) {
+    proto = proto[0];
+    var lowerProto = proto.toLowerCase();
+    if (!supportedProtocol[lowerProto]) {
+      throw new TypeError('Unsupported protocol');
+    }
+
+    proto = lowerProto;
+    rest = rest.slice(proto.length);
+
+    // Because there is a protocol, there has to be double slashes
+    // If there are not, the url is invalid
+    var slashes = rest.charCodeAt(0) === 47 &&
+                  rest.charCodeAt(1) === 47;
+    if (!slashes) {
+      throw new TypeError('Invalid URL: there is a protocol but no slashes');
+    } else {
+      rest = rest.slice(2);
+      // If there is a colon at the end of the protocol, remove it
+      if (lowerProto.charCodeAt(proto.length - 1) === 58) {
+        proto = proto.slice(0, proto.length - 1);
+      }
+    }
+
+    this.protocol = proto;
+  } else {
+    throw new TypeError('Invalid URL: no protocol');
+  }
+
+  this.host = rest;
+  if (this.host === '') {
+    throw new TypeError('Invalid Url: there is no host');
+  }
+
+  // Now we will check the host for authentication, port number
+  // and the hostname itself
+  var atSign = -1;
+  for (var i = 0; i < rest.length; ++i) {
+    switch (rest.charCodeAt(i)) {
+      case 9:
+      case 10:
+      case 13:
+      case 32:
+      case 34:
+      case 37:
+      case 39:
+      case 59:
+      case 60:
+      case 62:
+      case 92:
+      case 94:
+      case 96:
+      case 123:
+      case 124:
+      case 125:
+        // Characters that are never ever allowed in a hostname from RFC 2396
+        throw new TypeError('Invalid URL: prohibited character in the host');
```

Comment:
`dns` module will validate a passing hostname. Do we need to check if the given the hostname string meets the specification here?